language/move-prover/src/spec_translator.rs

```diff
@@ -1426,28 +1431,110 @@ impl<'env> SpecTranslator<'env> {
         );
     }
 
-    pub fn emit_global_invariants_for_memory(&self, assume: bool, memory: QualifiedId<StructId>) {
-        let env = self.global_env();
-        self.emit_global_invariants(assume, env.get_global_invariants_for_memory(memory))
+    pub fn emit_on_access_global_invariant_assumes(&self, memory: QualifiedId<StructId>) {
+        self.emit_global_invariants(
+            true,
+            self.get_effective_global_invariants(memory, GlobalInvariantContext::AssumeOnAccess),
+        );
+    }
+
+    pub fn emit_on_update_global_invariant_asserts(&self, memory: QualifiedId<StructId>) {
+        self.emit_global_invariants(
+            false,
+            self.get_effective_global_invariants(memory, GlobalInvariantContext::Assert),
+        );
     }
 
-    fn emit_global_invariants(&self, assume: bool, invariants: Vec<GlobalId>) {
+    /// Returns the set of effective global invariants for a given memory in a given usage
+    /// context (on top-level function entry, on memory access, and before and after a memory
+    /// update). This uses properties on the invariants and program options to determine the
+    /// effective set.
+    fn get_effective_global_invariants(
+        &self,
+        mem: QualifiedId<StructId>,
+        context: GlobalInvariantContext,
+    ) -> Vec<GlobalId> {
+        // All invariants which refer to this memory.
+        let all_invariants = self.global_env().get_global_invariants_for_memory(mem);
+
+        // A predicate determining whether the invariant has the property `isolated`. Such
+        // invariants are not used as assumptions for other verification steps.
+        let is_isolated = |id: &GlobalId| {
+            self.global_env()
+                .get_global_invariant(*id)
+                .and_then(|inv| {
+                    self.global_env()
+                        .is_property_true(&inv.properties, CONDITION_ISOLATED)
+                })
+                .unwrap_or(false)
+        };
+        let is_connected = |id: &GlobalId| !is_isolated(id);
+
+        // A predicate which determines whether this invariant depends on any memory which
+        // is part of the verification problem, that is touches a memory which is declared by
+        // a module which is verified.
+        let is_verified = |id: &GlobalId| {
+            self.global_env()
+                .get_global_invariant(*id)
+                .map(|inv| {
+                    inv.mem_usage
+                        .iter()
+                        .any(|used| !self.global_env().get_module(used.module_id).is_dependency())
+                })
+                .unwrap_or(true)
+        };
+
+        // Option which determines whether invariants should be assumed at access instead of
+        // on function entry. If invariants or assumed on access they my be included many times.
+        // On the other hand, they may be localized to the VC sub-problem at hand.
```

Comment:
I didn't understand the advantage that is being captured by "they may be localized to the VC sub-problem at hand".  You could either expand on the explanation here (or we could discuss 1-1).

Reply:
The assumption only happens at write, not at read. Without this (introduced 2-3 weeks ago) many things wont terminate.

Reply:
The name of this new option is assume_invariant_on_access but your comment above indicates that it applies to write accesses but not read accesses.  Might it be better to rename the option?

Reply:
Sorry, misread your comment. This is actually about something else.

This PR brings back an option to assume invariants "lazily" as we did a while ago, instead of eagerly at function entry. I was curious what is actually faster. Postponing invariant assumptions until they are needed may avoid having useless quantifiers which confuse Z3. However, benchmarking did not show an advantage, in fact a slight disadvantage. So this option is off by default, but the code should stay so we can play with this.